launcher/src/main/java/org/apache/spark/launcher/ChildThreadAppHandle.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.util.logging.Logger;
+
+public class ChildThreadAppHandle extends AbstractSparkAppHandle {
+  private static final Logger LOG = Logger.getLogger(ChildThreadAppHandle.class.getName());
+
+  private Thread childThread;
+
+  public ChildThreadAppHandle(String secret, LauncherServer server) {
+    super(server, secret);
+  }
+
+  @Override
+  public synchronized void kill() {
+    if (!disposed) {
+      disconnect();
+    }
+    if (childThread != null) {
+      try {
+        childThread.join(3000);
+      } catch (InterruptedException e) {
+        try {
+          childThread.interrupt();
+        } catch (Exception inner) {
+          LOG.info("Failed to stop Thread: " + inner.getMessage());
+        }
+      } finally {
+        childThread = null;
+      }
+    }
+  }
+
+  void setChildThread(Thread childThread) {
+    this.childThread = childThread;
+  }
+
+  protected void waitFor() throws InterruptedException {
+    if(this.childThread.isAlive()) {
+      this.childThread.wait();
```

Comment:
Why `wait` and not `join`? Who's doing a `notify` on the thread to wake up this call?